lib/internal/util/debuglog.js

```diff
@@ -115,7 +126,371 @@ function debuglog(set, cb) {
+  timesStore,
+  traceCategory,
+  implementation,
+  timerFlags,
+  logImpl,
+  logLabel = 'default',
+  traceLabel = undefined,
+) {
+  // Coerces everything other than Symbol to a string
+  logLabel = `${logLabel}`;
+
+  if (traceLabel !== undefined) {
+    traceLabel = `${traceLabel}`;
+  } else {
+    traceLabel = logLabel;
+  }
+
+  if ((timerFlags & kSkipLog) === 0) {
+    timeLogImpl(timesStore, implementation, logImpl, logLabel);
+  }
+
+  if ((timerFlags & kSkipTrace) === 0) {
+    traceLabel = safeTraceLabel(traceLabel);
+    trace(kTraceEnd, traceCategory, traceLabel, 0);
+  }
+
+  timesStore.delete(logLabel);
+}
+
+/**
+ * @param {SafeMap} timesStore
+ * @param {string} traceCategory
+ * @param {string} implementation
+ * @param {number} timerFlags
+ * @param {LogImpl} logImpl
+ * @param {string} logLabel
+ * @param {string} traceLabel
+ * @param {any[]} args
+ * @returns {void}
+ */
+function timeLog(
+  timesStore,
+  traceCategory,
+  implementation,
+  timerFlags,
+  logImpl,
+  logLabel = 'default',
+  traceLabel = undefined,
+  args,
+) {
+  // Coerces everything other than Symbol to a string
+  logLabel = `${logLabel}`;
+
+  if (traceLabel !== undefined) {
+    traceLabel = `${traceLabel}`;
+  } else {
+    traceLabel = logLabel;
+  }
+
+  if ((timerFlags & kSkipLog) === 0) {
+    timeLogImpl(timesStore, implementation, logImpl, logLabel, args);
+  }
+
+  if ((timerFlags & kSkipTrace) === 0) {
+    traceLabel = safeTraceLabel(traceLabel);
+    trace(kTraceInstant, traceCategory, traceLabel, 0);
+  }
+}
+
+/**
+ * @type {Record<string, SafeMap>}
+ */
+let tracesStores;
+
+/**
+ * @typedef {(logLabel: string, traceLabel?: string) => void} TimerStart
+ */
+
+/**
+ * @typedef {(logLabel: string, traceLabel?: string) => void} TimerEnd
+ */
+
+/**
+ * @typedef {(logLabel: string, traceLabel?: string, args?: any[]) => void} TimerLog
+ */
+
+/**
+ * Debuglog with time fns and support for trace
+ * @param {string} set
+ * @param {(startTimer: TimerStart, endTimer: TimerEnd, logTimer: TimerLog) => void} cb
+ * @returns {{startTimer: TimerStart, endTimer: TimerEnd, logTimer: TimerLog}}
+ */
+function debugWithTimer(set, cb) {
+  set = StringPrototypeToUpperCase(set);
+
+  if (tracesStores === undefined) {
+    tracesStores = { __proto__: null };
+  }
+
+  /**
+   * @type {LogImpl}
+   */
+  function logImpl(label, timeFormatted, args) {
+    const pid = process.pid;
+    const colors = { colors: lazyUtilColors().shouldColorize(process.stderr) };
+    const coloredPID = inspect(pid, colors);
+
+    if (args === undefined)
+      process.stderr.write(format('%s %s %s: %s\n', set, coloredPID, label, timeFormatted));
+    else
+      process.stderr.write(
+        format(
+          '%s %s %s: %s\n',
+          set,
+          coloredPID,
+          label,
+          timeFormatted,
+          ...new SafeArrayIterator(args),
+        ),
+      );
+  }
+
+  const kTraceCategory = `node,node.debuglog_${StringPrototypeToLowerCase(set)}`;
```

Comment:
Do we really need the `debuglog_` prefix though? It seems fine to just use the label directly as category.

Reply:
If we plan to use this helper in other places that already have tracing capabilities, maybe having this directly to `node.${set}` will be better since the user only needs to enable one category.

Initially, I wanted to have this completely separated to avoid collisions, but if we consider the above use-case, could make sense to remove the prefix.